docs/source/testing.rst

```diff
@@ -0,0 +1,218 @@
+
+
+GuiTestAssistant
+================
+
+.. note::
+
+   GuiTestAssistant is currently only available on Qt.
+
+:class:`GuiTestAssistant` is a mixin class intended to to augment
+:class:`python:unittest.TestCase`. In fact, it has
+:class:`~traits.testing.unittest_tools.UnittestTools` from traits as a base
+class and thus gives access to methods like
+:meth:`~traits.testing.unittest_tools.UnittestTools.assertTraitChanges`. See the
+`Traits Testing documentation <https://docs.enthought.com/traits/traits_user_manual/testing.html#testing>`_
+for more. 
+
+:class:`GuiTestAssistant` holds a reference to a :class:`pyface.gui.GUI` object
+(for api details see the interface :class:`~pyface.i_gui.IGUI`) which is what
+gives the low level access to the event loop. :class:`pyface.gui.GUI` provides methods such as
+:meth:`~pyface.i_gui.IGUI.start_event_loop`,
+:meth:`~pyface.i_gui.IGUI.stop_event_loop`,
+:meth:`~pyface.i_gui.IGUI.process_events`,
+:meth:`~pyface.i_gui.IGUI.invoke_later`, and
+:meth:`~pyface.i_gui.IGUI.set_trait_later`. This is accessible via the
+:attr:`gui` attribute on :class:`GuiTestAssistant`.
+
+What :class:`GuiTestAssistant` provides that is novel, is effectively more
+security to make sure your tests clean up after themselves.  For example,
+:class:`GuiTestAssistant` provides standard :meth:`setUp` and :meth:`tearDown`
+methods which try to clean up existing UI state and empty the event loop even
+if a test fails.  In addition, the methods typically have timouts so that the
+test will fail rather than blocking forever in the case something has gone
+wrong. Effectively, the class aims to remember to do the overhead to ensure
+your tests don't cause trouble, and at the same time give you the low level
+event loop access needed to write your GUI tests.
+
+
+This class provides the following methods (some of them being context managers):
+
+- :meth:`event_loop`
+  
+    Takes an integer ``repeat`` parameter and artificially replicates the event
+    loop by calling :meth:`sendPostedEvents` and :meth:`processEvents` ``repeat``
+    number of times.
+
+- :meth:`event_loop_until_condition`
+
+    Runs the real Qt event loop until the provided condition evaluates to True.
+
+- :meth:`event_loop_until_traits_change`
+
+    Run the real application event loop until a change notification for all of
+    the specified traits is received.
+
+- :meth:`event_loop_with_timeout`
+
+    Helper context manager to send all posted events to the event queue
+    and wait for them to be processed.
+
+    This differs from the `event_loop()` context manager in that it
+    starts the real event loop rather than emulating it with
+    ``QApplication.processEvents()``
+
+- :meth:`find_qt_widget`
+
+    Takes parameters ``start``, ``type_`` and ``test``. Recursively walks the Qt
+    widget tree from Qt widget ``start`` until it finds a widget of type ``type_``
+    (a QWidget subclass) that satisfies the provided ``test`` method.
```

Comment:
I would add a note here regd the finicky/crashy nature of `find_qt_widget` and point to traitsui UITester as an easier to use alternative.